test/interfaces/background_field_interface.jl

```diff
@@ -0,0 +1,231 @@
+    if l == zero(l)
+        return -cos(x)
+    elseif l == one(l)
+        # integral sin(x) e^(i x) dx = (i x)/2 - 1/4 e^(2 i x) + constant
+        return 0.5 * (1im * x - 0.5 * exp(2im * x))
+    else
+        # integral sin(x) e^(i k x) dx = (e^(i k x) (cos(x) - i k sin(x)))/(k^2 - 1) + constant
+        return exp(1im * x * l) * (cos(x) - 1im * l * sin(x)) / (l^2 - 1)
+    end
+end
+
+function _groundtruth_generic_spectrum(pol, l)
+    return _indefinite_integral(pol, rightendpoint(RND_DOMAIN), l) -
+           _indefinite_integral(pol, leftendpoint(RND_DOMAIN), l)
+end
+
+struct TestBGfield <: AbstractPulsedPlaneWaveField end
+
+QEDfields.reference_momentum(field::TestBGfield) = RND_MOM
+QEDfields.domain(::TestBGfield) = RND_DOMAIN
+QEDfields.phase_duration(::TestBGfield) = RND_DOMAIN_WIDTH
+QEDfields._envelope(::TestBGfield, x::Real) = one(x)
+
+struct TestBGfieldFAIL <: AbstractPulsedPlaneWaveField end
+
+@testset "hard interface" begin
+    test_field = TestBGfield()
+    @test reference_momentum(test_field) == RND_MOM
+    @test domain(test_field) == RND_DOMAIN
+    @test phase_duration(test_field) == RND_DOMAIN_WIDTH
+end
+
+@testset "failed interface" begin
+    test_field_FAIL = TestBGfieldFAIL()
+    @test_throws MethodError reference_momentum(test_field_FAIL)
+    @test_throws MethodError domain(test_field_FAIL)
+    @test_throws MethodError phase_duration(test_field_FAIL)
+end
+
+@testset "pulse envelope" begin
+    test_field = TestBGfield()
+
+    @testset "compute single" begin
+        rnd_phi = rand(RNG, RND_DOMAIN)
+        @test isapprox(
+            envelope(test_field, rnd_phi),
+            _groundtruth_envelope(rnd_phi),
+            atol=ATOL,
+            rtol=RTOL,
+        )
+        @test isapprox(
+            envelope(test_field, leftendpoint(RND_DOMAIN)),
+            _groundtruth_envelope(leftendpoint(RND_DOMAIN)),
+            atol=ATOL,
+            rtol=RTOL,
+        )
+        @test isapprox(
+            envelope(test_field, leftendpoint(RND_DOMAIN) - eps()),
+            zero(Float64),
+            atol=ATOL,
+            rtol=RTOL,
+        )
+        @test isapprox(
+            envelope(test_field, rightendpoint(RND_DOMAIN)),
+            _groundtruth_envelope(rightendpoint(RND_DOMAIN)),
+            atol=ATOL,
+            rtol=RTOL,
+        )
+        @test isapprox(
+            envelope(test_field, rightendpoint(RND_DOMAIN) + eps()),
+            zero(Float64),
+            atol=ATOL,
+            rtol=RTOL,
+        )
+    end
+    @testset "compute vector" begin
+        rnd_phis = rand(RNG, RND_DOMAIN, 2)
+        push!(rnd_phis, leftendpoint(RND_DOMAIN))
+        push!(rnd_phis, leftendpoint(RND_DOMAIN) - eps())
+        push!(rnd_phis, rightendpoint(RND_DOMAIN))
+        push!(rnd_phis, rightendpoint(RND_DOMAIN) + eps())
+
+        test_envelope_values = envelope(test_field, rnd_phis)
+
+        groundtruth_envelope_values = Vector{Float64}(undef, length(rnd_phis))
```

Comment:
Personally, I do not like the right hand side syntax for creating an uninitialized list. Does the list need to be uninitialized? Can't you just initialize it to `-1` or something? What is the advantage of an uninitialized list?

Reply:
ping @szabo137

Reply:
In this case, you are right because an uninitialized list is just a list of zeros, which are somewhat valid values for an envelope. This could be misleading in the later comparison. Therefore, I follow your suggestion and initialize the list with `-1.0` (see 73f7bfc). 

Generally, one uses this type of uninitialized list to allocate memory for a given element type and length of the list. For performance reasons, you usually want that instead of empty lists and pushing (see, e.g., [here](https://discourse.julialang.org/t/improving-my-mental-model-why-is-sizehint-so-much-slower-than-zeros/78776) ).